app/src/swap/handle_sign_transaction.c

```diff
@@ -0,0 +1,165 @@
+    const char *default_chain_id = "cosmoshub-4";
+    CHECK_PARSER_ERR(parser_getItem(ctx_parsed_tx, displayIdx, tmpKey, sizeof(tmpKey), tmpValue, sizeof(tmpValue), pageIdx, &pageCount))
+
+    // Check if chain_id is printed, expert fields
+    if (strcmp(tmpKey, "Chain ID") == 0) {
+        // For now allow only default chain id
+        if (strcmp(tmpValue, default_chain_id) != 0) {
+            ZEMU_LOGF(200, "Wrong Chain Id. ('%s', should be : '%s').\n", tmpValue, default_chain_id);
+            return parser_unexpected_error;
+        }
+        displayIdx += 5; // skip account_number, sequence, source_address, source_coins
+    } else {
+        displayIdx += 2; // skipsource_address, source_coins
+    }
+
+    // Check destination address
+    CHECK_PARSER_ERR(parser_getItem(ctx_parsed_tx, displayIdx, tmpKey, sizeof(tmpKey), tmpValue, sizeof(tmpValue), pageIdx, &pageCount))
+    if (strcmp(tmpKey, "Dest Address") != 0 || strcmp(tmpValue, G_swap_state.destination_address) != 0) {
+        ZEMU_LOGF(200, "Wrong swap tx destination address ('%s', should be : '%s').\n", tmpValue, G_swap_state.destination_address);
+        return parser_unexpected_error;
+    }
+
+    // Check destination coins
+    char tmp_amount[100] = {0};
+    zxerr_t zxerr = bytesAmountToStringBalance(G_swap_state.amount, G_swap_state.amount_length, tmp_amount, sizeof(tmp_amount));
+    if (zxerr != zxerr_ok) {
+        return parser_unexpected_error;
+    }
+    
+    displayIdx += 1;
+    CHECK_PARSER_ERR(parser_getItem(ctx_parsed_tx, displayIdx, tmpKey, sizeof(tmpKey), tmpValue, sizeof(tmpValue), pageIdx, &pageCount))
+    if (strcmp(tmpKey, "Dest Coins") != 0 || strcmp(tmp_amount, tmpValue) != 0) {
+        ZEMU_LOGF(200, "Wrong swap tx destination coins ('%s', should be : '%s').\n", tmpValue, tmp_amount);
+        return parser_unexpected_error;
+    }
+
+    // Check if memo is present
+    displayIdx += 1;
+    CHECK_PARSER_ERR(parser_getItem(ctx_parsed_tx, displayIdx, tmpKey, sizeof(tmpKey), tmpValue, sizeof(tmpValue), pageIdx, &pageCount))
+    if (strcmp(tmpKey, "Memo") == 0) {
+        if(strcmp(tmpValue, G_swap_state.memo) != 0) {
+            ZEMU_LOGF(200, "Wrong swap tx memo ('%s', should be : '%s').\n", tmpValue, G_swap_state.memo);
+            return parser_unexpected_error;
+        }
+
+        // Prepare for fees
+        displayIdx += 1;   
+        CHECK_PARSER_ERR(parser_getItem(ctx_parsed_tx, displayIdx, tmpKey, sizeof(tmpKey), tmpValue, sizeof(tmpValue), pageIdx, &pageCount))
+    }
+
+    //Check fees
+    zxerr = bytesAmountToStringBalance(G_swap_state.fees, G_swap_state.fees_length, tmp_amount, sizeof(tmp_amount));
+    if (zxerr != zxerr_ok) {
+        return parser_unexpected_error;
+    }
+
+    if (strcmp(tmpKey, "Fee") != 0 || strcmp(tmp_amount, tmpValue) != 0) {
+        ZEMU_LOGF(200, "Wrong swap tx fees ('%s', should be : '%s').\n",  tmpValue, tmp_amount);
+        return parser_unexpected_error;
+    }
+
```

Comment:
I'm not a cosmos expert at all so this is an open question
Are we ensuring that we can't be fooled into signing a more complex transaction with additional fields?
For example on your Polkadot application you check the number of fields of the TX